### PR TITLE
Update RBAC guide for plugin writers

### LINKED ISSUE #2463: Update plugin writers docs for RBAC
Write about roles and less Permissions. Remove any vestige of django-guardian.

### PR DESCRIPTION
**Pull request by mdellweg**
Explain the concept of roles, their relationship to permissions, how
they are associated with personas and objects and also remove any
mention of django-guardian that is to be removed in 3.20.

fixes #2463

### CODE REVIEW

File: docs/plugins/plugin-writer/concepts/rbac/access_policy.rst [hunk 1]
```diff
@@ -83,9 +83,7 @@ authorized.
 
     The ``admin`` user created on installations prior to RBAC being enabled has
     ``is_superuser=True``. Django assumes a superuser has any model-level permission even without it
-    being assigned. Additionally, django-guardian when checking object-level permissions defaults to
-    assuming the same although it is configurable. Generally, superusers are expected to bypass
-    authorization checks.
+    being assigned. Generally, superusers are expected to bypass authorization checks.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/access_policy.rst:86]:
```suggestion
    being assigned. Django's permission checking machinery assumes superusers bypass authorization checks.
```

I'm not sure if my suggestion is better, but the idea is to reduce the arbitrariness by identifying this is really a Django behavior.

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 2]
```diff
@@ -1,31 +1,26 @@
-Permissions
-===========
+Permissions and Roles
+=====================
 
-The permissions system provides a way to assign permissions to specific users and groups of users.
-The model driving this data is provided by ``django.contrib.auth.models.Permission``. Each
-``Permission`` has a name, describing it and can be associated with one or more users or groups.
+The permissions system provides a way to assign permissions to specific users and groups of users as part of roles.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:4]:
```suggestion
The permissions system provides a way to assign permissions to specific users and groups as part of roles.
```

Maybe better I'm not sure?

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 3]
```diff
@@ -1,31 +1,26 @@
...
+The models driving this data are ``django.contrib.auth.models.Permission`` and ``pulpcore.plugin.models.role.Role``. Each
+``Permission`` has a name, describing it and can be associated with one or more ``Role``.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:6]:
```suggestion
``Permission`` has a name, describing it and can be associated with one or more ``Role`` objects.
```

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 4]
```diff
@@ -40,34 +35,16 @@ Here's an example of the Permissions automatically created for the ``FileRemote`
...
 
 Defining Custom Permissions
 ---------------------------
 
-Any model can define a custom permission, and Django will automatically make a migration to add it
+Any model can define custom permissions, and Django will automatically make a migration to add it
 for you. See the `Django Custom Permissions Documentation <https://docs.djangoproject.com/en/2.2/
 topics/auth/customizing/#custom-permissions>`_ for more information on how to do that.
+In contrast to ``AccessPolicies`` and ``creation_hooks``, permissions can only be defined by the plugin writer.
+As a rule of thumb, permissions should be the atomic building blocks for roles and each action that can be performed on an object should have its own permission.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:47]: nitpick and I know it's just a draft, but in the final version plz line wrap this one.

**Comment by bmbouter** [comment]
This is all pretty good. What I really think would benefit this is a clear example, like one from pulp_file. If an example could be picked, e.g. FileRemote and then use that as the context to concretely talk about each thing I think it would do wonders to connect the abstract language (which is needed) to the concrete implementation of that (also needed).

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst (hunk 3 above)
Comment (mdellweg) [reply docs/plugins/plugin-writer/concepts/rbac/permissions.rst:6]: I'd rather not, as the `object` has some distinct meaning in the context of this file.

File: docs/plugins/plugin-writer/concepts/rbac/access_policy.rst [hunk 5]
```diff
@@ -235,3 +233,24 @@ different Permission check by declaring the ``permission_classes`` check. For ex
...
+-----------------------------
+
+drf-access-policy provides a feature to enable conditional checks to be globally available as their
+docs `describe here <https://rsinger86.github.io/ drf-access-policy/reusable_conditions/>`_. Pulp
+enables the ``reusable_conditions`` in its settings.py file, allowing a variety of condition
+checks to be globally available. Pulp enables this as follows:
+
+.. code-block:: python
+
+    DRF_ACCESS_POLICY = {"reusable_conditions": "pulpcore.app.global_access_conditions"}
```
Comment (ipanova) [inline docs/plugins/plugin-writer/concepts/rbac/access_policy.rst:250]: This needs update according to https://github.com/pulp/pulpcore/pull/2496/files

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 6]
```diff
@@ -91,31 +68,53 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
+        "file.fileremote_owner": [
+            "file.view_fileremote",
+            "file.change_fileremote",
+            "file.delete_fileremote",
+            "file.manage_roles_fileremote",
+        ],
+        "file.fileremote_viewer": ["file.view_fileremote"],
+    }
+
+Roles com in two flavors.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:113]:
```suggestion
Roles come in two flavors.
```

Comment (bmbouter) [reply docs/plugins/plugin-writer/concepts/rbac/permissions.rst:113]: If this is the topic sentence for the paragraph is should probably name the two types right here. Or perhaps the two types could each be their own paragraph. Either would read well to me, with a slight preference for one paragraph.

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 7]
```diff
@@ -91,31 +68,53 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
+First there are so called locked roles that are provided by plugins.
+Their name needs to be prefixed by the plugin ``app_label`` followed by a dot (see the example above).
+They can be seen, but not modified via the api, and are kept up to date with their definition in the plugin code.
+That way, plugins can ship default access policies that rely on those roles.
+The other flavor are user defined roles.
+These are managed and owned by the site admin via the Pulp API, and plugin code will not interfere with them.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:119]:
```suggestion
These are managed and via the Pulp API, and plugin code will not interfere with them.
```

It's hard to say who is really in charge of this stuff, all we really know is it's done through an API.

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 8]
```diff
@@ -91,31 +68,53 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
+Users can opt to use the the provided locked roles or roll their own.
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:120]: homophone pun!

Comment (mdellweg) [reply docs/plugins/plugin-writer/concepts/rbac/permissions.rst:120]:
```suggestion
Users can opt to use the provided locked roles or roll their own.
```

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 9]
```diff
@@ -40,34 +37,17 @@ Here's an example of the Permissions automatically created for the ``FileRemote`
...
-
-
 .. _defining_custom_permissions:
 
 Defining Custom Permissions
 ---------------------------
 
-Any model can define a custom permission, and Django will automatically make a migration to add it
+Any model can define custom permissions, and Django will automatically make a migration to add it
 for you. See the `Django Custom Permissions Documentation <https://docs.djangoproject.com/en/2.2/
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:46]:
```suggestion
for you. See the `Django Custom Permissions Documentation <https://docs.djangoproject.com/en/3.2/
```

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 10]
```diff
@@ -91,31 +71,56 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
 
-drf-access-policy provides a feature to enable conditional checks to be globalls available as their
-docs `describe here <https://rsinger86.github.io/ drf-access-policy/reusable_conditions/>`_. Pulp
-enables the ``reusable_conditions`` in its settings.py file, allowing a variety of condition
-checks to be globally available. Pulp enables this as follows:
+Roles
+-----
 
-.. code-block:: python
+``Roles`` are basically sets of ``Permissions`` and in Pulp, users and groups should receive their
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:89]:
```suggestion
``Roles`` are basically sets of ``Permissions``, and in Pulp, users and groups should receive their
```

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 11]
```diff
@@ -91,31 +71,56 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
+``Permissions`` exclusively via role assignments. Typical roles are ``owner`` for an object with all
+the permissions to view modify and delete the object, or ``viewer`` limited to see the object. To
+scope the reach of the permissions in a role, these role are assigned to ``Users`` or ``Groups``
+either on the model-level or the object-level.
 
-    DRF_ACCESS_POLICY = {"reusable_conditions": "pulpcore.app.global_access_conditions"}
+:Model-Level: A role is associated to a user or group for access to a specific model, but not an
+   instance of that model. This allows you to express concepts like "Hilde can administer all
```
Comment (bmbouter) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:96]: Nice use of `Hilde` as a specific user persona.

**Review by bmbouter — APPROVED** [review]
Very tiny optional comments left. Thanks for doing this!

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 12]
```diff
@@ -1,31 +1,28 @@
...
+Model Permissions
+-----------------
 
-.. _model_level_permissions:
-
-Model-Level Permissions
------------------------
-
-By default, each model receives four permissions:
+``Permissions`` in Django (or should i say Pulp?) are tied to models and usually map to certain
```
Comment (ipanova) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:16]: maybe just say :" ``Permissions`` are tied to models.."

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst [hunk 13]
```diff
@@ -91,31 +71,56 @@ Here's an example of adding a permission like this for ``FileRepository``:
...
+            "file.manage_roles_fileremote",
+        ],
+        "file.fileremote_viewer": ["file.view_fileremote"],
+    }
+
+Roles come in two flavors, locked and user-defined. First there are so called locked roles that are
+provided by plugins. Their name needs to be prefixed by the plugin ``app_label`` followed by a dot
+(see the example above). They can be seen, but not modified via the api, and are kept up to date
+with their definition in the plugin code. That way, plugins can ship default access policies that
+rely on those roles. The other flavor are user defined roles. These are managed and via the Pulp
```
Comment (gerrod3) [inline docs/plugins/plugin-writer/concepts/rbac/permissions.rst:124]:
```suggestion
rely on those roles. The other flavor is user defined roles. These are managed via the Pulp
```

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst (hunk 11 above)
Comment (mdellweg) [reply docs/plugins/plugin-writer/concepts/rbac/permissions.rst:96]: nih

File: docs/plugins/plugin-writer/concepts/rbac/permissions.rst (hunk 12 above)
Comment (mdellweg) [reply docs/plugins/plugin-writer/concepts/rbac/permissions.rst:16]:
```suggestion
``Permissions`` in Django are tied to models and usually map to certain
```

Let's keep Django as pointed out in another comment, this is not our fault.

**Review by ipanova — APPROVED** [review]
Thank you!

**Pull request merged**